javalib/src/main/scala/java/util/LinkedList.scala

```diff
@@ -0,0 +1,356 @@
+      }
+    }
+
+    if (head.getValue().isEmpty)
+      check(last)
+    else
+      check(head)
+  }
+
+  override def size(): Int =
+    _size
+
+  override def add(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  override def remove(o: Any): Boolean = {
+    val boxed = Box(o.asInstanceOf[E])
+    val iter = listIterator()
+    var changed = false
+    while (iter.hasNext()) {
+      if (Box(iter.next()).equals(boxed)) {
+        iter.remove()
+        changed = true
+      }
+    }
+
+    changed
+  }
+
+  override def addAll(c: Collection[_ <: E]): Boolean = {
+    val iter = c.iterator
+    var changed = iter.hasNext()
+    while (iter.hasNext())
+      addLast(iter.next())
+
+    changed
+  }
+
+  override def clear(): Unit = {
+    head = new Node(None)
+    last = new Node(None)
+    _size = 0
+  }
+
+  override def get(index: Int): E = {
+    checkIndexInBounds(index)
+
+    throw new Exception("TBD")
+  }
+
+  override def set(index: Int, element: E): E = {
+    add(index, element)
+    remove(index+1)
+  }
+
+  override def add(index: Int, element: E): Unit = {
+    checkIndexOnBounds(index)
+    
+    throw new Exception("TBD")
+  }
+
+  override def remove(index: Int): E = {
+    checkIndexInBounds(index)
+    
+    throw new Exception("TBD")
+  }
+
+  override def indexOf(o: Any): Int = {
+    throw new Exception("TBD")
+  }
+
+  override def lastIndexOf(o: Any): Int = {
+    throw new Exception("TBD")
+  }
+
+  def peek(): E =
+    peekFirst()
+
+  def element(): E =
+    getFirst()
+
+  def poll(): E =
+    pollFirst()
+
+  def remove(): E =
+    remove(0)
+
+  def offer(e: E): Boolean =
+    offerLast(e)
+
+  def offerFirst(e: E): Boolean = {
+    addFirst(e)
+    true
+  }
+
+  def offerLast(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  def peekFirst(): E = {
+    throw new Exception("TBD")
+  }
+
+  def peekLast(): E = {
+    throw new Exception("TBD")
+  }
+
+  def pollFirst(): E = {
+    val ret = peekFirst()
+    if (ret == null) null.asInstanceOf[E] else {
+      removeFirst()
+      ret
+    }
+  }
+
+  def pollLast(): E = {
+    val ret = peekLast()
+    if (ret == null) null.asInstanceOf[E] else {
+      removeLast()
+      ret
+    }
+  }
+
+  def push(e: E): Unit =
+    addFirst(e)
+
+  def pop(): E =
+    removeFirst()
+
+  def removeFirstOccurrence(o: Any): Boolean = {
+    val iter = iterator()
+    var changed = false
+    while (iter.hasNext() && !changed)
+      if (iter.next() === o) {
+        iter.remove()
+        changed = true
+      }
+
+    changed
+  }
+
+  def removeLastOccurrence(o: Any): Boolean = {
+    val iter = descendingIterator()
+    var changed = false
+    while (iter.hasNext() && !changed)
+      if (iter.next() === o) {
+        iter.remove()
+        changed = true
+      }
+
+    changed
+  }
+
+  override def listIterator(index: Int): ListIterator[E] = {
+    checkIndexOnBounds(index)
+    new RandomAccessListIterator(self, index, 0, size)
```

Comment:
This is clearly not class that has `RandomAccess` and therefore `RandomAccessListIterator ` will be horribly inefficient. You will have to define a custom `ListIterator`. It should keep the reference to the current node (the one that has the value that `next` will return), the rest should be quite simple.

Tip: When you create this iterator you can enter the list through the `head` or through `last`. You should allays use the one that is closer to `index`. If you do this, you will be able to remove `override def add(e: E) ...` because the default implementation will reach the end in constant time (there might be others like that).